### PR TITLE
Assignment 2 - text adventure game - group 4

### PR DESCRIPTION
**Pull request by Edward2k**
Our assignment for assignment2.

### CODE REVIEW

File: docs/assignment1.md [hunk 1]
```diff
@@ -1,62 +1,73 @@
...
+To interact with the VUORK world, the user will use text commands in the CLI to give instructions. The instruction will be in the form “<action> <items…*>”. We will create a parser which will transform that instruction into something our system can understand: If the user types in a command, describing what he is doing/what he wants to do, he/she will get a response from the computer what she/he is now able to see, what has changed or if the move is not valid. Furthermore, instead of having a boring CLI as the realm, we want to add a few Graphical elements into the game. Examples include the players inventory, the current areas explored etc...
+
+With the game briefly explained, we must now consider how to develop this idea in a way that is easily extendable, maintainable and comprehensible. By using the Object-Oriented programming paradigm, we aim to create a system that is easily modifiable, expandable and intuitive for programmers to understand. To help achieve this goal, we will be making use of UML diagrams. 
+
+   				   Figure 1: A use case diagram of VUORK
+
+![VuORK use case diagram](https://i.postimg.cc/SRBV1RM5/Screen-Shot-2020-03-01-at-6-40-11-PM.png)
```
Comment (Ferch42) [inline docs/assignment1.md:22]: Incorrect use of use case diagram. It should represent use cases and should not have this directed arrows

File: docs/assignment1.md [hunk 2]
```diff
@@ -1,62 +1,73 @@
...
...ine interface. It is, thus, intrisically important to have a parser that will understand some simple semantics of langauge. For example, a command such as *"north"* should be equivilant to *walk north*. <br> After careful consideration, we have decided that our parser must have the input in the form of *<action\> <item(s)/objects(s) [proposition]...\*>*. An item may be preceeded by a proposition. For example, *"Put cheese in fridge"* or *"Attack bear with knife"*. The action put, will let the parser know that there will be 2 items, interupted by some proposition. Other actions, however, may exclude a propostion. An example is *"Take broom"* or *"Move north"*.<br>Emposing this structure will make it possible to parse natural language to something our system can understand. If the user, for any reason, excludes a second item in a 2-part command (such as attack), then the system will reply with *"What do you want to attack the <object\> with?"*, if the item is ommited. Or *"What do you want to attack?"*, if an object is omitted. If an invalid propositon is used, for example *"Attack monkey on knife"*, the parser should reply, *"I do not understand how to Attack monkey on knife"*<br> |
```
Comment (Ferch42) [inline docs/assignment1.md:37]: {feedback from assignment 1 not properly addressed} It is still unclear what are the possible actions described by the commands. Try defining precisely what are the available commands, what happens when they are invoked and how the game is affected by them. Try making them simple as simple as possible.

File: docs/assignment1.md [hunk 3]
```diff
@@ -1,62 +1,73 @@
...
 
-Don’t forget to use links to external URLs (e.g., the direct link to the Fantasy soccer league you are getting inspiration from, the link to the original video game you are getting inspiration from, etc.), if applicable. 
+After a plethora of discourse with the team, we have decided that our game, titled: **VUORK: Saving Thilo from the Chamber of Secrets**, will take place at the Vrije Universiteit Amsterdam. The user assumes the role of a newly enrolled student (user may choose the name) on his way to a 9AM lecture of Computer Programming. Albeit new, when the user arrives at the main campus of the VU, he notices something is wrong… This is where the game begins. 
 
-Maximum number of words for this section: 1000
+After a brief introduction into the game, the user will discover he is alone at the VU, and his lecturer, Thilo, is lost in the Chamber of Secrets. He will then have to gather the puzzle pieces, each with a clue indicating where the next piece of the puzzle is, to find the chamber of secrets; where Thilo is held captive.
```
Comment (Ferch42) [inline docs/assignment1.md:14]: Please provide a detailed explanation of the gameplay, game objectives and game mechanics.

File: docs/assignment1.md [hunk 4]
```diff
@@ -1,62 +1,73 @@
...
+| F4   | Game-info  | The player can get the statistics with the command *stats*. The statistics include, time being played, current score (see F7). Furthermore, information about the current gamestate can be achieved with the use of special commands. See F3 for more info on these. |
+| F5   | Map        | A list that keeps track of the areas already visited. Quite self explanatory. **Bonus:** Include an option to add a graphic for each item which can then be displayed as a backpack. |
```
Comment (Ferch42) [inline docs/assignment1.md:41]: Bonus describing wrong graphic display. it does not make sense for a backpack to be displayed in the map

File: docs/assignment1.md [hunk 5]
```diff
@@ -1,62 +1,73 @@
...
+| F6   | Backpack   | A list that keeps track of the items that have been picked up already. Also quite self explanatory. **Bonus:** Include an option to add a graphic for each item which can then be displayed as a backpack. |
+| F7   | Score      | The score will be an integer value which will be a numeric representation as to how well the player will is playing. It is calculated by summing the move with the total damage taken. Lower is better, much like in golf. |
```
Comment (Ferch42) [inline docs/assignment1.md:43]: Describe what is the objective of the game and in what manner the score is related to the game

File: docs/assignment1.md [hunk 6]
```diff
@@ -1,62 +1,73 @@
...
+| F8   | Moves      | An integer that shows the number of moves at the top of the game as part of the text-based GUI. Always updated after an action. This is used immediatly display to the user how many steps the user has taken (and thus, how efficiently he is completing the game in). |
```
Comment (Ferch42) [inline docs/assignment1.md:44]: Why is this feature important?

File: docs/assignment1.md [hunk 7]
```diff
@@ -1,62 +1,73 @@
...
+| QR2 | Check valid command | Correctness | If the user gives a command that is legal, but is not applicable at the current state, the sytem should ALWAYS respond, in some form, that the input was correct, just not allowed in the current state of the game. |
+| QR3 | Quick     responses to user input     | Responsiveness | The system must always respond to any input. It shall, thus, not be possible for the user to be able to input 2 seperate commands without a response from the system. The system should respond immediatly. |
+| QR4 | Extensible world | Maintainability | The code should be well structured and organized; and the universe should not be closed. It should be an open story plot so more levels can be added later. Readable code will make it easier to implement such. |
+| QR5 | User friendly | Usability | An introduction should explain how to play the game and how to give commands. The syntax shall be intuitive for any player. The help of text-based GUI will help give a visual display of current inventory, current layout of the explored map etc. |
```
Comment (Ferch42) [inline docs/assignment1.md:59]: Features described here, not quality requirements

File: docs/assignment1.md [hunk 8]
```diff
@@ -1,62 +1,73 @@
...
+| QR6 | Always respond to user inputs | Reliability | All commands shall be idempotent; If a player drops an item, that item shall be dropped. |
```
Comment (Ferch42) [inline docs/assignment1.md:60]: Quality requirement: Idempotent commands

File: docs/assignment2.md [hunk 9]
```diff
@@ -1 +1,409 @@
...
+
+Number of words in this document: 8000
+
+**Format**: establish formatting conventions when  describing your models in this document. For example, you style the name of each class is *Italics*, whereas the attributes, operations, and  associations as <u>underlined</u> text, objects are in italic, etc.
+
+###
+
+### Implemented feature
+
+| ID   | Short name | Description                                                  |
```
Comment (Ferch42) [inline docs/assignment2.md:11]: As suggested in the feedback above, it is necessary clarifications about the gameplay structure, game objectives and so forth.

File: docs/assignment2.md [hunk 10]
```diff
@@ -1 +1,409 @@
...
+
+This chapter contains the specification of the UML class diagram of your system, together with a textual description of all its elements.
+
+```
+Class Diagram of VuORK
+```
+
+> Figure representing the Class Diagram of VuORK (Diagram 1) **DESCRIPTIVE**
+
+![Class Diagram of VuORK](https://i.postimg.cc/DZX98JfB/Class-diagram-Class-diagram.png)
```
Comment (Ferch42) [inline docs/assignment2.md:38]: Abstract class and multiplicity of the associations are not conformed with standard uml notation. Duplication between an attribute and a relationship (with the same name) - only one is enough (relationships preferred)

File: docs/assignment1.md (hunk 2 above)
Comment (Edward2k) [reply docs/assignment1.md:37]: thank you for your feedback. However, as stated in the documentation, we did not give a list of possible actions because this should be easily created by a 'creater' who can parse his/her JSON file with Areas, and Items. Each Item would later be defined to have certain traits (also in the JSON files)

File: docs/assignment2.md (hunk 10 above)
Comment (Edward2k) [reply docs/assignment2.md:38]: Thank you again :). Just to be clear, what do you mean 'only one is enough'. Looking at the <<Item>> abstract class, do we need only one line to Area/Player?

Also, I do not understand what is wrong with the multiplicity. Looking at the slides, it seems (to me), that if fits into the provided UML guidelines. 

Also, What did you think of the structure of the diagram?

File: docs/assignment2.md [hunk 11]
```diff
@@ -1 +1,409 @@
...
+
+<u>*getID()*</u> returns the ID of the item
+
+Item has a composition relation to (exclusivley) *Area* or a *Player*. This is because an Item can not both exist outside of the *Player's backpack* and in the surrounding area or vice versa. Furthermore, an Item can not exist without an *Area* or *Player's backpack* to exist in. 
+
+<h5 id="BasicItem">BasicItem</h5>
+
+Singelton is a subtype of <a href="#Item">Item</a> that a player can interact with. It is the most basic type of item. It inherits properties from Item, such as the Name and ID and their consequent getter functions. 
+
+An Item can be interracted with, or used to interract. To elaborate, imagine the command ***Attack chair with sword***. Here, the action attack has 2 clauses: the dependent and independent. This means that BasicItems need a way to know what can be done to it (ie be attacked), and what it can be used to do (ie attack). In this case, the BasicItem chair would have the string *Attack*, in its *UsedTo*.
```
Comment (Ferch42) [inline docs/assignment2.md:58]: Here a glimpse of the game mechanics is presented. Please provide a more detailed explanation of how the game mechanics should work in the description of the system (assignment 1)

File: docs/assignment2.md [hunk 12]
```diff
@@ -1 +1,409 @@
...
+
+<u>movePlayer(coord: Coordinate): Bool</u> will move a player iff the condition *Map.isValidMove(coord: Coordinate)* is true.
+
+*<u>getCurrentPosition(): Coordinate</u>* returns the value of <u>coord</u>
+
+*<u>getInstruction</u>()*, will return the ask *getInstruction* from *Parser* assosciated to the *Player*. 
+
+*<u>setName(String): Void</u>* sets the value of <u>name</u> to String.
+
+The relations of Player are a little complex. A player must contain a *Parser* and a *Printer* (equally, these 2 must can not exist without a *Player*). The *Player* needs these 2 classes to manage I/O individually in a multiplayer scenario. Much like in *Area*, a *Player* object may contain a reference to an *Item*, iff this *Item* does not already exist in an *Area*
```
Comment (Ferch42) [inline docs/assignment2.md]: Consider creating a game_engine layer from which both players can interact at the same time. A couple of possible problems, such as syncronization and consistency of the game world, might be mitigated this way

File: docs/assignment2.md [hunk 13]
```diff
@@ -1 +1,409 @@
...
+
+If the command is help: a help message will be printed to the user.
+
+If the command is look: the description of the *Area*, in that the *Player* is currently in, will be printed to the user. 
+
+<h5>State-Machine Diagram for Player Movement</h5>
+
+> Figure representing the state machine diagram of the class Map (Diagram 4) **DESCRIPTIVE**
+
+[![State-Machine-Diagram-Page-2.png](https://i.postimg.cc/FHgLtyV2/State-Machine-Diagram-Page-2-Page-2.png)](https://postimg.cc/PLZ2PpXG)
```
Comment (Ferch42) [inline docs/assignment2.md:316]: Events for the decision node are not specified. The name of the states correspond errouneously to event conditions.

File: docs/assignment2.md [hunk 14]
```diff
@@ -1 +1,409 @@
...
+
+
+#### Command Processing
+
+
+> Figure representing the sequence of events for processing commands after the game has been initialised(Diagram 7) 
+
+
+[![Sequence-diagram-valid-command-valid-command.png](https://i.postimg.cc/d0qj4c4x/Sequence-diagram-valid-command-valid-command.png)](https://postimg.cc/RWDtZyr7)
```
Comment (Ferch42) [inline docs/assignment2.md:352]: alt fragment for generating coordinate not necessary, as it does not include any special behaviour. In the last alt fragment the map returns a description (not a call to get description from the printer)

**Comment by Ferch42** [comment]
Merge approved. Please resolve the merge conflicts and merge branch

**Comment by Ferch42** [comment]
Some mistakes in some of the diagrams. Most important criticism is that it was hard visualizing the idea of the project and then map the diagrams and models to the intended result. Please provide a clear and fleshed out description about the gameplay, game mechanics and objective (in addition maybe with an example).

File: docs/assignment1.md (hunk 2 above)
Comment (Ferch42) [reply docs/assignment1.md:37]: Hi edward. As I stated on my review, from the descriptions provided is very difficult to get an idea of the overall intention of the project, the game structure, objectives, etc. From that it was also hard mapping the requirements to the models. If that is the case, please provide a more detailed explanation about the realtionship between the creator and the game mechanics

File: docs/assignment2.md (hunk 10 above)
Comment (Ferch42) [reply docs/assignment2.md:38]: The multiplicity of the associations should not be enclosed with "[]", and the multiplicity, when displayed should appear on both ends of the association. Abstract classes should have "{abstract}". When you make an association between classes you do not have to display it as an attribute also (see slide 35 from the second lecture).

Comment (Edward2k) [reply docs/assignment2.md:38]: Ah yes the "[]" was my bad. No clue why I added them. However, looking at slide 38 of Lecture 2, there is multiplicity given only at 1 end? That is what I followedas a guide to my mutiplicity. I ommitted the other multiplicty when there was a composite aggregation, because that means that Class A would not exist without Class B. 
Is this implicit description not valid UML?

File: docs/assignment1.md [hunk 15]
```diff
@@ -1,62 +1,73 @@
...
-Write a short description of your version of the system you are going to design and implement in this course. 
-Clearly specify which are the key aspects of your system, such as:
-- customizations/extensions w.r.t. the project track, 
-- main type of user(s)
-- overall idea about how it works
-- ...
+Inspiration: [Original Zork game](https://textadventures.co.uk/games/play/5zyoqrsugeopel3ffhz_vq)
 
-Be creative here!
+For our software project, we have decided to recreate the game Zork. Zork is one of the first Command Line Interface (CLI), role playing games (RPG). In Zork, a user assumes the role of a protagonist stuck in a post-apocalyptic world. To win the game, the protagonist must find all 20 treasures, and place them in the trophy case. Our recreation of Zork will remodel the game play with: a new setting, a new goal, new commands etcetera.
```
Comment (PSvasquez) [inline docs/assignment1.md:10]: I would have to agree with @Ferch42 that there could be a more detailed explanation of the gameplay in this section and that it does seem a bit bare bones. As someone who isn't that familiar with the original Zork, it would be useful to include how players will interact with the interface with more specificity.

File: docs/assignment1.md [hunk 16]
```diff
@@ -1,62 +1,73 @@
...
 
-Maximum number of words for this section: 1000
+After a brief introduction into the game, the user will discover he is alone at the VU, and his lecturer, Thilo, is lost in the Chamber of Secrets. He will then have to gather the puzzle pieces, each with a clue indicating where the next piece of the puzzle is, to find the chamber of secrets; where Thilo is held captive. 
 
-## Features
-Author(s): `name of the team member(s) responsible for this section`
+To interact with the VUORK world, the user will use text commands in the CLI to give instructions. The instruction will be in the form “<action> <items…*>”. We will create a parser which will transform that instruction into something our system can understand: If the user types in a command, describing what he is doing/what he wants to do, he/she will get a response from the computer what she/he is now able to see, what has changed or if the move is not valid. Furthermore, instead of having a boring CLI as the realm, we want to add a few Graphical elements into the game. Examples include the players inventory, the current areas explored etc...
```
Comment (PSvasquez) [inline docs/assignment1.md:16]: The description here is a bit hard to follow/ read coherently with the constant changes from "he is doing..." to "he/she will get a response..." to simplify your text description and make it more readable try using gender-neutral they/them.

File: src/main/java/Parser.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+		Scanner in = new Scanner(System.in);
+		return in.nextLine();
+	}
+
+	private String[] getLineArray() {
+		Scanner parseScanner = new Scanner(System.in);
+		String line = parseScanner.nextLine();
+		line = line.toLowerCase();
+		Scanner lineScanner = new Scanner(line);
+		String[] command = new String[MAX_WORDS_PER_COMMAND];
```
Comment (GitOutOfMySwamp) [inline src/main/java/Parser.java:21]: My understanding is that the remainder of the input more than 4 words is ignored? If this is the case, the system should throw some sort of exception to notify the user, otherwise they may not realize a part of their command has been ignored.

File: src/main/java/Player.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+    private Parser parser;
+    private Printer printer;
+
+    Player (Coordinate coord) {
+        this.coord = coord;
+        this.backpack = new Item[]{};
+        score = 0;
+        moves = 0;
+        health = 100;
```
Comment (GitOutOfMySwamp) [inline src/main/java/Player.java:18]: Where do you edit the user's score, moves and health?

File: src/main/java/Game.java [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	public static void validateCommand(Instruction command, Player player) {
+		String action = command.getAction();
+		//Movement shortcut
+		if (isDirection(action)) {handleMove(action, player); return;}
+		switch(action) {
+			case "move":
+				handleMove(command.getItems().get(0), player);
+				break;
```
Comment (GitOutOfMySwamp) [inline src/main/java/Game.java:45]: Validating the player that has given the command seems unnecessary in a single player game.

File: src/main/java/Game.java [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+		while(true) {
+			Instruction currCommand = player.getInstruction(); //Blocking command waiting for input
+			validateCommand(currCommand, player); //validateCommand
+		}
+	}
+
+	public static void validateCommand(Instruction command, Player player) {
+		String action = command.getAction();
+		//Movement shortcut
+		if (isDirection(action)) {handleMove(action, player); return;}
```
Comment (GitOutOfMySwamp) [inline src/main/java/Game.java:41]: Redundant to have handleMove in the switch statement as well

**Review by GitOutOfMySwamp — COMMENTED** [review]
Very well written code

**Comment by tomsiebring** [comment]
Very intuitive code with good overview for reader. Separating functions clearly amongst the classes helps with this, good job.

Some general comments that catched my eye:
- I notice many instantiations of objects throughout the code. Furthermore, functionalities have been divided up into many individual lines. E.g. getLineArray() could perhaps be given a more compact shape.
- Good/appropriate use of getters and setters for Player class


_private static final int CHARS_PER_LINE = 100;
    void output(String output) {
        boolean newLine = false;
        for (int i = 0; i < output.length(); i++) {
            if (i % CHARS_PER_LINE == 0 && i != 0) {newLine = true;}
            if (newLine && output.charAt(i) == ' ') {
                System.out.println();
                i++;
                newLine = false;
            }
            System.out.print(output.charAt(i));
        }
        System.out.println();
    }_
- Printer class could be made more readable for fellow reviewers
- Nice instructions 


_private void initDemoMap() {
        entryPoint = new Coordinate(2, 1);
        mapSize = 3;
        map = new Area[mapSize][mapSize];
//        |  N |    N      |  Eating Area      |
//        |  N | LOBBY     |  Cafeteria        |
//        |  N | Entrance  |        N          |
        map[2][1] = new Area("Entrance", "It is a quiet Friday morning and you find yourself directly infront of the " +
                "of the VU main building. For some reason, you can not leave this area, and are forced to move north of where you are now.");
        map[1][1] = new Area("Lobby", "Everything is quieter than usual in the Lobby. The lights are off and it is, thus, darker than usual. You do, however, see a light to your East.");
        map[1][2] = new Area("Cafeteria", "The cafeteria is, like the lobby, empty. You do, however, hear some noise... you just can't tell from where :O");
        map[0][2] = new Area("Eating area", "You found the noise! It is coming from some rats eating a rotten sandwich. Also worth noting that, if your made it here, it is the end of the DEMO map :p");_

- Try to prevent code copying as much as possible, e.g. in this case a wrapper or other strategy might be more elegant

File: src/main/java/Parser.java [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+		return command;
+	}
+
+	protected Instruction getInstruction() {
+		String[] parts = getLineArray(); 
+		while (parts[0] == null) {
+			parts = getLineArray();
+		}
+		String act = parts[ACTION_WORD_POSITION]; //Get first element for action
+		parts = Arrays.copyOfRange(parts, START_OF_ITEMS, MAX_WORDS_PER_COMMAND); //Each command can have max 4 words.
```
Comment (PSvasquez) [inline src/main/java/Parser.java:35]: nice use of comments and overall very clean code :)

File: docs/assignment1.md [hunk 22]
```diff
@@ -1,62 +1,73 @@
...
...with the user: Let the user know what is going on. The narrator will describe the *field of view* of the player at any given time when: Moving between rooms, the command info is received. We define *field of view* as what all objects and properties of the immediate area around the user. That is, the area the user is currently in. For example, if the player is in a room that is dark, there is only an entrace to the west, and in this area, there lies a fridge, broom and coffee machine. The game would narrate *"You are in a dark room. The the west you see an opening to another area. In this room, you see a fridge, broom and coffee machine"*.  <br>There are also other commands available, which will have a special naration. For example, when picking up and object, the narrator will reply with a simple confirmation, simply repeating the action. If the action fails, the system should reply accordingly. <br>The command *diagnose* will trigger the narrator to list the health of the player. For example, you are wounded on your left hand, 80% health remaining<br>The command *inventory* will list all of the contents, without any specific ordering, the player currently has in their posession. |
```
Comment (PSvasquez) [inline docs/assignment1.md:39]: I like the use of specific examples and this makes the features easier to visualize, continue to do more of this!

File: docs/assignment1.md (hunk 15 above)
Comment (Edward2k) [reply docs/assignment1.md:10]: Thanks man! But I do not understand what is so hard to understand :p It's obvious I am doing something wrong. Would it be worthwhile to completly rewrite it? As in, was it really unclear and hard to visaulize?

File: docs/assignment2.md (hunk 10 above)
Comment (Ferch42) [reply docs/assignment2.md:38]: I checked the slide 38 and it shows the multiplicity at all ends of the association, as a ternary association. It is still necessary to show the multiplicity even in a composite aggregation (see slide 47 of lecture 2)
